--- cdparanoia-III-10.2/interface/interface.c.org	2024-12-27 11:34:19.000000000 +0100
+++ cdparanoia-III-10.2/interface/interface.c	2024-12-27 11:34:19.000000000 +0100
@@ -39,9 +39,9 @@
     if(d->drive_model)free(d->drive_model);
     if(d->cdda_fd!=-1)close(d->cdda_fd);
     if(d->ioctl_fd!=-1 && d->ioctl_fd!=d->cdda_fd)close(d->ioctl_fd);
-    if(d->private){
-      if(d->private->sg_hd)free(d->private->sg_hd);
-      free(d->private);
+    if(d->private_data){
+      if(d->private_data->sg_hd)free(d->private_data->sg_hd);
+      free(d->private_data);
     }
 
     free(d);
@@ -127,7 +127,7 @@
 	}
       }	
     }
-    if(ms)*ms=d->private->last_milliseconds;
+    if(ms)*ms=d->private_data->last_milliseconds;
     return(sectors);
   }
   
